### PR TITLE
Postprocess hardening

### PR DESCRIPTION
**Pull request by freddyaboulton**
## Description

Ensures only files located in the current working directory or tempdir are able to be moved to the cache

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

File: gradio/routes.py [hunk 1]
```diff
@@ -553,44 +553,22 @@ async def file(path_or_url: str, request: fastapi.Request):
...
+                allowed_paths=blocks.allowed_paths
+                + [app.uploaded_file_dir, utils.get_cache_folder()]
+                + _StaticFiles.all_paths,
+                file_sets=blocks.temp_file_sets,
+            )
+            if not allowed:
                 raise HTTPException(403, f"File not allowed: {path_or_url}.")
 
-            if not abs_path.exists():
-                raise HTTPException(404, f"File not found: {path_or_url}.")
```
Comment (freddyaboulton) [inline gradio/routes.py:592]: According to the unit tests, we actually don't raise 404 if the file does not exist

File: gradio/blocks.py [hunk 2]
```diff
@@ -1910,6 +1909,10 @@ async def process_api(
         state_ids_to_track, hashed_values = self.get_state_ids_to_track(block_fn, state)
         changed_state_ids = []
 
+        from gradio.context import LocalContext
+
+        LocalContext.blocks.set(self)
```
Comment (abidlabs) [inline gradio/blocks.py:1914]: Interesting, why do we set it here and not in the init of Blocks?

File: gradio/processing_utils.py [hunk 3]
```diff
@@ -462,6 +457,55 @@ def _move_to_cache(d: dict):
     return client_utils.traverse(data, _move_to_cache, client_utils.is_file_obj)
 
 
+def _check_allowed(path: str | Path, check_in_upload_folder: bool):
+    blocks = LocalContext.blocks.get()
```
Comment (abidlabs) [inline gradio/processing_utils.py:461]: is not sufficient to use Context.root_block?

File: gradio/processing_utils.py [hunk 4]
```diff
@@ -462,6 +457,55 @@ def _move_to_cache(d: dict):
...
+    if blocks is None or not blocks.is_running:
+        return
+
+    import tempfile
```
Comment (abidlabs) [inline gradio/processing_utils.py:465]: suggest putting this import at top of file

File: gradio/processing_utils.py [hunk 5]
```diff
@@ -462,6 +457,55 @@ def _move_to_cache(d: dict):
...
+        if reason == "in_blocklist":
+            msg += f"it is located in one of the blocked_paths ({', '.join(blocks.blocked_paths)})."
+        elif check_in_upload_folder:
+            msg += "it was not uploaded by a user."
+        else:
+            msg += "it was not created by the application or it is not "
+            msg += "located in either the current working directory or your system's temp directory. "
+            msg += "To fix this error, please ensure your function returns files located in either "
+            msg += f"the current working directory ({os.getcwd()}), your system's temp directory ({tempfile.gettempdir()}) "
+            msg += f"or add {str(abs_path.parent)} to the allowed_paths parameter."
```
Comment (abidlabs) [inline gradio/processing_utils.py:494]: I would add something like, "if you'd like to specifically allow this file to be served, you can add it to the `allowed_paths` parameter of `launch()`"

File: gradio/processing_utils.py [hunk 6]
```diff
@@ -462,6 +457,55 @@ def _move_to_cache(d: dict):
...
+    if (
+        utils.is_in_or_equal(abs_path, os.getcwd())
+        and abs_path.name.startswith(".")
+        and not any(
+            is_in_or_equal(path, allowed_path) for allowed_path in blocks.allowed_paths
+        )
+    ):
+        raise InvalidPathError(
+            "Dotfiles located in the temporary directory cannot be moved to the cache for security reasons. "
+            "Please check whether this file is valid and rename it if so."
```
Comment (abidlabs) [inline gradio/processing_utils.py:505]: same^

File: test/test_utils.py [hunk 7]
```diff
@@ -414,6 +424,52 @@ def test_is_in_or_equal_fuzzer(path_1, path_2):
...
 
 
+@settings(derandomize=os.getenv("CI") is not None)
+@given(
+    path=create_path_string(),
+    blocked_paths=create_path_list(),
+    allowed_paths=create_path_list(),
+    file_sets=st.lists(create_path_set(), min_size=0, max_size=3),
+)
+def test_is_allowed_file_fuzzer(
```
Comment (abidlabs) [inline test/test_utils.py:434]: Cool idea to use the fuzzer here. But if I'm understanding correctly, randomly generating paths is very, very unlikely to produce any overlaps between the parameters, right? So this is actually unlikely to test what happens if a `path` is in `blocked_paths` or `allowed_paths` or `file_sets`? Perhaps part of the path should be manually specified or the values should be chosen from a set with smaller cardinality

File: gradio/routes.py [hunk 8]
```diff
@@ -547,44 +547,22 @@ async def file(path_or_url: str, request: fastapi.Request):
...
-            in_blocklist = any(
-                utils.is_in_or_equal(abs_path, blocked_path)
-                for blocked_path in blocks.blocked_paths
-            )
-
-            is_dir = abs_path.is_dir()
-
-            if is_dir or in_blocklist:
+            if abs_path.is_dir() or not abs_path.exists():
                 raise HTTPException(403, f"File not allowed: {path_or_url}.")
```
Comment (abidlabs) [inline gradio/routes.py:551]: nit: it would be better to set an `allowed` flag here, and raise a single 403 later, after the `utils.is_allowed_file` has completed so that malicious attackers cannot using timing attacks to infer whether a particular filepath even exists (that itself can be insecure in certain settings)

**Comment by abidlabs** [comment]
Note: I believe the `gr.Code` component is still vulnerable to attacks since in `postprocess()`, we simply read the file and return its contents to users. I think we just should just deprecate letting `gr.Code` taking in a tuple value

**Comment by abidlabs** [comment]
Left some comments but overall this works great. Good stuff @freddyaboulton

**Review by abidlabs — APPROVED** [review]

File: gradio/processing_utils.py (hunk 3 above)
Comment (freddyaboulton) [reply gradio/processing_utils.py:461]: I think we set Context.root_block to `None` when we exit the Blocks context and launch

File: gradio/blocks.py (hunk 2 above)
Comment (freddyaboulton) [reply gradio/blocks.py:1914]: I thought this would be more "future proof" as it's the most up to date version of the Blocks before any sort of processing happens. But I don't think we need this actually.

File: gradio/routes.py (hunk 8 above)
Comment (freddyaboulton) [reply gradio/routes.py:551]: Yea good call.

File: gradio/processing_utils.py (hunk 3 above)
Comment (abidlabs) [reply gradio/processing_utils.py:461]: Ah I see

File: test/test_utils.py (hunk 7 above)
Comment (freddyaboulton) [reply test/test_utils.py:434]: I greatly reduced the cardinality of the allowed possible paths set and also added some corner cases manually in a different test.

File: gradio/processing_utils.py (hunk 5 above)
Comment (freddyaboulton) [reply gradio/processing_utils.py:494]: done

File: gradio/processing_utils.py (hunk 4 above)
Comment (freddyaboulton) [reply gradio/processing_utils.py:465]: Done

File: gradio/blocks.py (hunk 2 above)
Comment (freddyaboulton) [reply gradio/blocks.py:1911]: We need this to be done after init for `is_running` to be set. I think actually this is a good place for it.
